### ISSUE TITLE
Undefined reference

### ISSUE THREAD

**Issue by yashamon** [latex-lsp/texlab#923]
This issue appeared recently. Probably last release. All citations are now marked undefined reference. The location of bib is standard: ~/texmf/bibtex/bib
The document compiles without issues with all references intact. 

Platform: Win11 release, neovim 9.1

**Comment by pfoerster** [comment]
@yashamon Thanks for the report! Does `kpsewhich` find the bibliography and does `texlab` (check if `texlab` reports a document link for the `\bibliography`/`\addbibresource` command)? As a test, can you try adding the directory to the `BIBINPUTS` environment variable, please?

**Comment by firubat** [comment]
@pfoerster This bug happens to me too. Neovim 0.9.2 on linux mint.
`kpsewhich` finds the bibliography.
Can you explain how to do the check in `texlab`?

**Comment by yashamon** [comment]
I thought I also replied. In my case kpsewhich also finds the bib file. We are only using the LSP server so I am not sure how to troubleshoot further.

**Comment by pfoerster** [comment]
@yashamon @firubat 
Can you try triggering a "goto definition" on the `\addbibresource` or `\bibliography` command? If it does not open the file, then the server does not find the file.

Can you also try the `BIBINPUTS` test, please (note that you cannot use `~` inside the variable)?

> As a test, can you try adding the directory to the BIBINPUTS environment variable, please?

**Comment by firubat** [comment]
@pfoerster So seems that if the bib file is in the same folder as the tex file, it is recognized by texlab. But if the bib file is in a different folder in the BIBINPUTS path (which is the way I like to do it), it is not recognized (the goto definition won't open the file in this case).

**Comment by yashamon** [comment]
Same for me, the issue goes away if the bib file is local.

**Comment by pfoerster** [comment]
@yashamon @firubat Can you try b2a49d0bfc5bde1d56c26fa8afdf55c5fe5c34cb, please? With this changeset, the server should be able to find the bibliography.

**Comment by naibaf0** [comment]
I have the same error even if the actual scenario seems different and just tried the new commit. It does find a bibliography file in the tex distribution and uses it now, however even in the case where a local file with the same name is present.

in my case I do have a local file `bibliography.bib` in my project which should be used and when running with this patch, goto definition (of the bibliography) will now offer to go to `/usr/share/doc/texlive/bibtex/newcastle-bst/bibliography.bib` in addition to my local file. It also offers autocompletion for citations from both files.

So I am guessing there is still some logic missing to prefer a local file over the one from the tex distribution tree.

**Comment by firubat** [comment]
> Can you try [b2a49d0](https://github.com/latex-lsp/texlab/commit/b2a49d0bfc5bde1d56c26fa8afdf55c5fe5c34cb), please?

I must admit that I don't know how to do that, I'm using texlab in neovim via Mason, and I don't know how to tell it to use a specific commit which is outside the releases cycle...

**Comment by naibaf0** [comment]
@firubat for me mason saves the binary for texlab in `~/.local/share/nvim/mason/packages/texlab`. I think that is the default, but not completely sure.
If you build texlab on the the specified commit locally, you can copy the newly built binary into the location where mason saves it and it will be used in neovim.

**Comment by pfoerster** [comment]
@naibaf0 Thanks for trying out the commit. 

> So I am guessing there is still some logic missing to prefer a local file over the one from the tex distribution tree.

This should be fixed now. I will publish a new release soon.

**Comment by firubat** [comment]
Hi @pfoerster , for some reason this is still an issue for me (current version 5.11.0)

**Comment by yashamon** [comment]
Is it already in the release version?

I have no way to compile commits, well I would prefer not to.

**Comment by pfoerster** [comment]
@yashamon 

> Is it already in the release version?


Yes, this changeset is included with [v5.9.0](https://github.com/latex-lsp/texlab/blob/master/CHANGELOG.md#590---2023-08-06).

@firubat Is the missing bibliography contained in your home directory?

**Comment by yashamon** [comment]
In that case I am still experiencing the issue. My bib is in the supposedly
standard global directory. If bib file is moved to local directory the
issue goes away.

**Comment by firubat** [comment]
@pfoerster It's in ~/Dropbox.

**Comment by firubat** [comment]
I tried to put it also in /usr/share/texmf and /usr/local/share/texmf but to no avail. Only if it's in the same directory as the tex file.

**Comment by Zeta611** [comment]
I have this issue when there are (at least) Korean characters in the path.

I have a `.bib` file in a local directory where the `.tex` file lives in.
If the directory contains Korean characters, the LSP shows `Undefined reference` error, but when I move the directory to a  path without Korean characters, the error goes away.

This probably means the directory is not properly handled for CJK characters.

**Comment by andrasbarany** [comment]
I am also having this issue with version 5.12.3 in both `nvim` and `nvim-qt`. If I copy or link the bibliography file to the local directory, the error disappears. `BIBINPUTS` is set to the folder containing the `.bib` file but is apparently ignored.

I'm using my distributions `texlab` binary, but changing to the one installed using `:LspInstall` in nvim doesn't help either.

Here's the output of `:LspInfo`

```
 Press q or <Esc> to close this window. Press <Tab> to view server doc.
 
 Language client log: /home/ba/.local/state/nvim/lsp.log
 Detected filetype:   tex
 
 1 client(s) attached to this buffer: 
 
 Client: texlab (id: 1, bufnr: [1])
 	filetypes:       tex, plaintex, bib
 	autostart:       true
 	root directory:  /home/ba
 	cmd:             /usr/bin/texlab
 
 Configured servers list: texlab, marksman
 ```

**Comment by lucasvreis** [comment]
Perhaps it's related to the tilde expansion? 
```latex
\addbibresource{~/Zotero/bibs/all.bib}
```
wasn't working for me, but
```latex
\addbibresource{/home/lucas/Zotero/bibs/all.bib}
```
does.

**Comment by juntuu** [comment]
tl;dr Percent encoding in paths may have different casing of hex characters (a-f or A-F), which leads to mismatches later.

---

> I have this issue when there are (at least) Korean characters in the path.
> 
> I have a `.bib` file in a local directory where the `.tex` file lives in. If the directory contains Korean characters, the LSP shows `Undefined reference` error, but when I move the directory to a path without Korean characters, the error goes away.
> 
> This probably means the directory is not properly handled for CJK characters.

I also noticed this issue with some characters in the path. (using neovim as the LSP client)

I did some investigation and found out that (at least part of) the cause was the percent encoding of URIs.

The LSP messages from the client contain URIs, which were percent encoded with **lowercase** hex. However, the URI for loaded file is constructed from a Rust `Path`, see: https://github.com/latex-lsp/texlab/blob/3d70feb03dc7d2cd9e11a80f5a766ec84699dbd7/crates/base-db/src/workspace.rs#L95

That `from_file_path` method uses **uppercase** hex for percent encoding, which causes the URIs to not match later.

Another place where the bibliography path is constructed is: https://github.com/latex-lsp/texlab/blob/3d70feb03dc7d2cd9e11a80f5a766ec84699dbd7/crates/base-db/src/deps/graph.rs#L162

This joins the relative path with the directory component, which may come from the LSP client. Thus the joined path may have different casing for the hex in percent encoding, depending on the LSP client.

I think the first is loading the bib file as part of the workspace, and the latter is resolving the path given to `addbibresource` command. So the two URIs for the same file may end up being different.

<details>
<summary>
(Relative paths seem to work, when the encoded part is within it)
</summary>

After this I also noticed that traversing back the encoded parts of the path "worked" again. (Which kind of makes sense.)

Under path `/home/me/ä/`, where `ä` will be encoded.

Doesn't work
```tex
\addbibresource{references.bib}
\cite{test}
```

"Works"
```tex
\addbibresource{../ä/references.bib}
\cite{test}
```
</details>

---

Solutions?

I'm not familiar with the project or the code, so can't say what would make sense.

Could that be part of `normalize_uri`? https://github.com/latex-lsp/texlab/blob/3d70feb03dc7d2cd9e11a80f5a766ec84699dbd7/crates/texlab/src/util.rs#L9

**Comment by pfoerster** [comment]
@juntuu Thanks for digging into this. I have created #1148, which extends `normalize_uri`.

**Comment by nolanking90** [comment]
I'm having trouble duplicating this issue.  When I have a remote bib file, I get the undefined reference diagnostic when I first open the file.  The diagnostic goes away if I modify the document and texlab reparses it a second time. There is no initial diagnostic when I move the .bib file to the project root directory and open the tex file. 

Can someone with this issue confirm that the diagnostic remains after you modify and save the file?

**Comment by firubat** [comment]
@nolanking90 I can confirm this diagnostic appears also after I do changes and save (and compile) the file again.

**Comment by nolanking90** [comment]
> [@nolanking90](https://github.com/nolanking90) I can confirm this diagnostic appears also after I do changes and save (and compile) the file again.

Can you give me a minimal example that reproduces the issue? e.g. a main.tex with one reference, one citation in a bib file, and specify the paths of the bib and main.tex file? The output of `echo $BIBINPUTS` (from the shell command line) and the output of `:LspInfo` (in command mode in vim, while you have main.tex open).

I'm sorry to ask for all this boilerplate info again, but it's it sounds like there may be (or were) more than one bug being discussed in this issue.

**Comment by firubat** [comment]
> The output of `echo $BIBINPUTS` (from the shell command line)

OK, this request has shed some light for me. When I ran  `echo $BIBINPUTS` I got nothing. I have set the BIBINPUTS path through texmf.cnf, not in my shell, so that's no surprise.
I actually see that @pfoerster mentioned it right away:

> As a test, can you try adding the directory to the `BIBINPUTS` environment variable, please?

but I think I didn't understand the terminology, I thought that is what I did, while now I understand he probably meant the shell environment variable, not the texmf one. I'm sorry about the confusion!

So anyway, if I put in my `.bashrc` the line 
```
export BIBINPUTS=$TEXMFDOTDIR$TEXMF/bibtex/bib//:/my/custom/path
```
then the diagnostic goes away. So this is a workaround solution (I say workaround as I think `texlab` should be able to get that path from the TEXMF variable, rather than requiring me to add it in another place, but correct me if I'm wrong).

And if you want to reproduce the error, you'd probably need to remove BIBINPUTS from your shell, and set it only in texmf.cnf.

**Comment by yashamon** [comment]
I am still getting the error. I am on windows in nushell: 
`echo $env.BIBINPUTS
C:/Users/yasha/texmf/bibtex/bib` 
This directory contains the bibliography as expected.

**Comment by SdotVdot** [comment]
For me, (MACTEX), it only works if the reference (.bib) file is local.

**Comment by jdreaver** [comment]
I was also hitting this issue. I have my .tex file in a subdirectory, and it contains `\addbibresource{../references.bib}`. When I add a `.texlabroot` to the subdirectory containing my .tex file, my undefined reference errors disappear. (I discovered this while trying to figure out why `latex-lsp-build` wasn't working.)

**Comment by sstac1** [comment]
For me (texlive+neovim) the issue is still present (5.26). It works only if .bib is local. Has anyone managed to find some solution?
